src/emulator/gxm/src/texture_cache.cpp

```diff
@@ -0,0 +1,235 @@
+        case SCE_GXM_TEXTURE_BASE_FORMAT_X8S8S8U8:
+        case SCE_GXM_TEXTURE_BASE_FORMAT_X8U24:
+        case SCE_GXM_TEXTURE_BASE_FORMAT_U32:
+        case SCE_GXM_TEXTURE_BASE_FORMAT_S32:
+        case SCE_GXM_TEXTURE_BASE_FORMAT_SE5M9M9M9:
+        case SCE_GXM_TEXTURE_BASE_FORMAT_F11F11F10:
+            return 32;
+        case SCE_GXM_TEXTURE_BASE_FORMAT_F16F16F16F16:
+        case SCE_GXM_TEXTURE_BASE_FORMAT_U16U16U16U16:
+        case SCE_GXM_TEXTURE_BASE_FORMAT_S16S16S16S16:
+        case SCE_GXM_TEXTURE_BASE_FORMAT_F32F32:
+        case SCE_GXM_TEXTURE_BASE_FORMAT_U32U32:
+            return 64;
+        case SCE_GXM_TEXTURE_BASE_FORMAT_PVRT2BPP:
+            return 2;
+        case SCE_GXM_TEXTURE_BASE_FORMAT_PVRT4BPP:
+            return 4;
+        case SCE_GXM_TEXTURE_BASE_FORMAT_PVRTII2BPP:
+            return 2;
+        case SCE_GXM_TEXTURE_BASE_FORMAT_PVRTII4BPP:
+            return 4;
+        case SCE_GXM_TEXTURE_BASE_FORMAT_UBC1:
+            return 4;
+        case SCE_GXM_TEXTURE_BASE_FORMAT_UBC2:
+        case SCE_GXM_TEXTURE_BASE_FORMAT_UBC3:
+            return 8;
+        case SCE_GXM_TEXTURE_BASE_FORMAT_YUV420P2:
+        case SCE_GXM_TEXTURE_BASE_FORMAT_YUV420P3:
+        case SCE_GXM_TEXTURE_BASE_FORMAT_YUV422:
+            return 16; // NOTE: I'm not sure this is right.
+        case SCE_GXM_TEXTURE_BASE_FORMAT_P4:
+            return 4;
+        case SCE_GXM_TEXTURE_BASE_FORMAT_P8:
+            return 8;
+        case SCE_GXM_TEXTURE_BASE_FORMAT_U8U8U8:
+        case SCE_GXM_TEXTURE_BASE_FORMAT_S8S8S8:
+            return 24;
+        case SCE_GXM_TEXTURE_BASE_FORMAT_U2F10F10F10:
+            return 32;
+    }
+    
+    return 0;
+}
+
+static TextureCacheHash hash_data(const void *data, size_t size) {
+    const uint8_t *const begin = static_cast<const uint8_t *>(data);
+    const uint8_t *const end = begin + size;
+    return std::accumulate(begin, end, TextureCacheHash(0));
+}
+
+static TextureCacheHash hash_texture_data(const SceGxmTexture &texture, const MemState &mem) {
+    GXM_PROFILE(__func__);
+    
+    const SceGxmTextureFormat format = texture::get_format(&texture);
+    const SceGxmTextureBaseFormat base_format = texture::get_base_format(format);
+    const size_t width = texture::get_width(&texture);
+    const size_t height = texture::get_height(&texture);
+    const size_t stride = (width + 7) & ~7; // NOTE: This is correct only with linear textures.
+    const size_t bpp = bits_per_pixel(base_format);
+    const size_t size = (bpp * stride * height) / 8;
+    const Ptr<const void> data(texture.data_addr << 2);
+    
+    return hash_data(data.get(mem), size);
+}
+
+static const bool operator==(const SceGxmTexture &a, const SceGxmTexture &b) {
+    return memcmp(&a, &b, sizeof(a)) == 0;
+}
+
+static size_t find_lru(const TextureCacheTimestamps &timestamps, TextureCacheTimestamp current_time) {
+    GXM_PROFILE(__func__);
+    
+    uint64_t oldest_age = current_time - timestamps.front();
```

Comment:
What't the point of including `current_time` at all when calculating least recently used texture?

Considering `current_time` will be bigger than any given timestamp, isn't it enough to simply search for the entry with the smallest timestamp?

Reply:
It was to handle the edge case where the timestamp wraps. It's 64-bit so that _shouldn't_ happen, but you never know.